### PR TITLE
ingest nodemanager publishing perf improvement

### PR DESCRIPTION
**Pull request by jaym**
Optimize nodes_tags 
- Created an index on nodes_tags
- Missing (node_id,tag_id) pairs are added in a single statement

Parallelize nodemanager publisher for ingest-service
By default, we'll have 2 publishers:
```toml
[ingest.v1.sys.service]
number_of_nodemanager_publishers = 2
```

We can disable publishing to nodemanager:
```toml
[ingest.v1.sys.service]
number_of_nodemanager_publishers = -1
````

### CODE REVIEW

File: components/ingest-service/server/chef.go [hunk 1]
```diff
@@ -101,7 +101,7 @@ func (s *ChefIngestServer) ProcessChefAction(ctx context.Context, action *chef.A
 		defer close(errc)
 
 		var err error
-		if errQ := s.chefActionPipeline.Run(ctx, action, errc); err != nil {
+		if errQ := s.chefActionPipeline.Run(ctx, action, errc); errQ != nil {
```
Comment (jaym) [inline components/ingest-service/server/chef.go:104]: This was a deadlock/memory leak. Goroutines would sit around forever if the chef action queue filled up

Comment (stevendanna) [reply components/ingest-service/server/chef.go:104]: 💭 Could this be most of the problem you've been chasing for the customer?

Nice find.

Comment (jaym) [reply components/ingest-service/server/chef.go:104]: i did not see this in their core dump, but was able to consistently hit it on my machine when I did
`chef_load_nodes 2000`

File: components/ingest-service/pipeline/publisher/node_manager_publisher.go [hunk 2]
```diff
@@ -2,51 +2,70 @@ package publisher
...
+	"github.com/sirupsen/logrus"
 	log "github.com/sirupsen/logrus"
 )
 
-func BuildNodeManagerPublisher(nodeManagerClient manager.NodeManagerServiceClient) message.ChefRunPipe {
+func BuildNodeManagerPublisher(nodeManagerClient manager.NodeManagerServiceClient, numPublishers int) message.ChefRunPipe {
 	return func(in <-chan message.ChefRun) <-chan message.ChefRun {
-		return nodeManagerPublisher(in, nodeManagerClient)
+		if numPublishers <= 0 {
+			logrus.Info("Publishing to nodemanager is disabled")
```
Comment (vjeffrey) [inline components/ingest-service/pipeline/publisher/node_manager_publisher.go:20]: ooh, nice!

File: components/ingest-service/pipeline/publisher/node_manager_publisher.go [hunk 3]
```diff
@@ -2,51 +2,70 @@ package publisher
...
+				start := time.Now()
+				_, err = nodeManagerClient.ProcessNode(ctx, nodeMetadata)
+				if err != nil {
+					log.Errorf("unable to send info about node %s to node manager", msg.Node.NodeName)
+				}
+				dur := time.Since(start)
+				log.WithFields(log.Fields{
+					"message_id":  msg.ID,
+					"buffer_size": len(out),
+					"dur":         dur,
```
Comment (vjeffrey) [inline components/ingest-service/pipeline/publisher/node_manager_publisher.go:60]: 🎉

File: components/nodemanager-service/pgdb/tags.go [hunk 4]
```diff
@@ -13,8 +14,19 @@ FROM tags
...
+INSERT INTO nodes_tags(node_id, tag_id) 
+	(
+		SELECT $1, tid 
+		FROM 
+			unnest($2::TEXT[]) tid 
+		LEFT JOIN nodes_tags ON 
+			nodes_tags.tag_id=tid AND 
+			node_id=$1 
+		WHERE nodes_tags.node_id IS NULL
+	)
```
Comment (vjeffrey) [inline components/nodemanager-service/pgdb/tags.go:29]: nit: semicolon

File: components/nodemanager-service/pgdb/tags.go [hunk 5]
```diff
@@ -66,39 +79,7 @@ func (trans *DBTrans) tagExists(tag *common.Kv) (string, error) {
...
-		if !exists {
-			link := NodeTag{
-				NodeID: nodeID,
-				TagID:  tagID,
-			}
-			links = append(links, &link)
-		}
-	}
-	return trans.Insert(links...)
+	_, err := trans.Exec(sqlInsertNodeTag, nodeID, pq.Array(tagIDs))
```
Comment (vjeffrey) [inline components/nodemanager-service/pgdb/tags.go:83]: nice!

**Pull request merged**